### PR TITLE
feat: update build process

### PR DESCRIPTION
**Pull request by matthewkeil**
**Motivation**
Had some issues with compilation of rust bindings for `hashtree-js` which is a node.js wrapper around this lib.  The rust bindings did not build for Windows

**Inclusions**
- Fix `build.rs` to compile on Windows
- Update `Makefile`
  * Make platform selection more robust so its not necessary to pass in which `CC` for default situations
  * Allow for passing `CC` for cross-compilation situations
  * unify `objarm` and `objx86` into `OBJ_LIST` dependent on platform
- Add `Dockerfile` for compiling and testing rust bindings on `linux-arm64`
- Update workflow.yaml to test for more cases of both the base lib and the rust bindings.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -2,76 +2,99 @@ name: test
...
+          --platform linux/arm64 \
+          -t ${{ matrix.settings.target }} \
+          .
     - name: install rustup
+      if: ${{ !matrix.settings.use-docker }}
       run: |
         curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs > rustup-init.sh
         sh rustup-init.sh -y --default-toolchain none
-        rustup target add ${{ matrix.target }}
+        rustup target add ${{ matrix.settings.target }}
```
Comment (matthewkeil) [inline .github/workflows/test.yml:92]: As a note, this step can be replaced with an action

```yaml
      - name: Install
        uses: dtolnay/rust-toolchain@stable
        if: ${{ !matrix.settings.use-docker }}
        with:
          toolchain: stable
          targets: ${{ matrix.settings.target }}
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -2,76 +2,99 @@ name: test
...
+            target: aarch64-unknown-linux-gnu
+            runner: ubuntu-latest
+            use-docker: true
+          - name: Windows x86
+            target: x86_64-pc-windows-msvc
+            runner: windows-latest
+          # - name: Windows arm64
+          #   target: aarch64-pc-windows-msvc
+          #   runner: windows-latest
+          #   use-docker: true
```
Comment (matthewkeil) [inline .github/workflows/test.yml:70]: I had a hard time finding a container that was easy to setup for building Windows on arm64.  I suppose if someone wants in the future should be easy enough to turn on but ran out of time and just commented this case out.

**Comment by matthewkeil** [comment]
I also got x86 musl building natively and with rust, but not sure if you want to include that.  I had trouble with arm but also didnt try super hard because its sort of a minority architecture.  Could be turned on but wasnt sure of the implications relative to the assembly in the lib.  I think they are compatible and its just the build tooling and abi layout that changes. Honestly, I'm not super familiar with musl is why I removed it.

https://github.com/matthewkeil/hashtree/actions/runs/9272182080/job/25509333449

Screenshot 2024-05-28 at 10 41 29 PM

**Comment by arnetheduck** [comment]
are you using the rust bindings in js?

**Comment by matthewkeil** [comment]
> are you using the rust bindings in js?

Yes @arnetheduck.  In https://github.com/ChainSafe/hashtree-js/pull/3

Are using a submodule from the branch this PR is built from but will move to the crate once its published

**Review by prestonvanloon — APPROVED** [review]

**Comment by arnetheduck** [comment]
(sorry for off-topic)

@matthewkeil ah curious - I would have thought that the C files are both easier to use and "thinner" in terms of compiler support that's needed to get to the finish line - what motivates the use of rust in this case?

**Comment by matthewkeil** [comment]
> (sorry for off-topic)
> 
> @matthewkeil ah curious - I would have thought that the C files are both easier to use and "thinner" in terms of compiler support that's needed to get to the finish line - what motivates the use of rust in this case?

Our bindings were built with `napi-rs` in rust for this project, as a trial, to see how that library did from a DX perspective. Building JS binding with c++ is a bear of a job. Big picture, if we can move to rust we think it'll be better (at least we hope).

**Review by potuz — APPROVED** [review]
Thanks so much for this one, after offline discussions this looks good to me

**Pull request merged**